drivers/counter/counter_max32_timer.c

```diff
@@ -0,0 +1,345 @@
+/*
+ * Copyright (c) 2024 Analog Devices, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT adi_max32_counter
+
+#include <zephyr/drivers/counter.h>
+#include <zephyr/drivers/clock_control/adi_max32_clock_control.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/irq.h>
+
+#include <wrap_max32_tmr.h>
+
+/** MAX32 MCUs does not have multiple channel */
+#define MAX32_TIMER_CH 1U
+
+struct max32_tmr_data {
+	counter_top_callback_t top_callback;
+	void *top_user_data;
+	uint32_t guard_period;
+};
+
+struct max32_tmr_ch_data {
+	counter_alarm_callback_t callback;
+	void *user_data;
+};
+
+struct max32_tmr_config {
+	struct counter_config_info info;
+	struct max32_tmr_ch_data *ch_data;
+	mxc_tmr_regs_t *regs;
+	const struct device *clock;
+	struct max32_perclk perclk;
+	int clock_source;
+	int prescaler;
+	void (*irq_func)(const struct device *dev);
+	bool wakeup_source;
+};
+
+static int api_start(const struct device *dev)
+{
+	const struct max32_tmr_config *cfg = dev->config;
+
+	Wrap_MXC_TMR_EnableInt(cfg->regs);
+	MXC_TMR_Start(cfg->regs);
+
+	return 0;
+}
+
+static int api_stop(const struct device *dev)
+{
+	const struct max32_tmr_config *cfg = dev->config;
+
+	Wrap_MXC_TMR_DisableInt(cfg->regs);
+	MXC_TMR_Stop(cfg->regs);
+
+	return 0;
+}
+
+static int api_get_value(const struct device *dev, uint32_t *ticks)
+{
+	const struct max32_tmr_config *cfg = dev->config;
+
+	*ticks = MXC_TMR_GetCount(cfg->regs);
+	return 0;
+}
+
+static int api_set_top_value(const struct device *dev, const struct counter_top_cfg *counter_cfg)
+{
+	const struct max32_tmr_config *cfg = dev->config;
+
+	if (counter_cfg->ticks == 0) {
+		return -EINVAL;
+	}
+
+	if (counter_cfg->ticks != cfg->info.max_top_value) {
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static uint32_t api_get_pending_int(const struct device *dev)
+{
+	const struct max32_tmr_config *cfg = dev->config;
+
+	return Wrap_MXC_TMR_GetPendingInt(cfg->regs);
+}
+
+static uint32_t api_get_top_value(const struct device *dev)
+{
+	const struct max32_tmr_config *cfg = dev->config;
+
+	return cfg->info.max_top_value;
```

Comment:
Is this value unmodifiable?

Reply:
Not need to be modified, see https://github.com/analogdevicesinc/zephyr/blob/add-MAX32xxx-Counter-Timer/drivers/counter/counter_max32_timer.c#L321